include/lsst/partition/ConfigStore.h

```diff
@@ -0,0 +1,290 @@
+    static std::vector<T> convert(nlohmann::json const& param, std::string const& path) {
+        std::string const context = "ConfigStore::Value<std::vector<T>>::convert: ";
+        if (!param.is_array()) {
+            throw lsst::partition::ConfigTypeError(context + "parameter: '" + path + "' is not an array.");
+        }
+        std::vector<T> v;
+        for (auto const& e: param) {
+            v.push_back(Value<T>::convert(e, path));
+        }
+        return v;
+    }
+};
+
+
+/**
+ * Template specialization of the utility class Value to allow interpreting single
+ * character strings as values of the 'char' type.
+ */
+template <>
+class Value<char> {
+public:
+    static char convert(nlohmann::json const& param, std::string const& path) {
+        std::string const context = "ConfigStore::Value<char>::convert: ";
+        if (param.is_string()) {
+            std::string const s = Value<std::string>::convert(param, path);
+            if (s.size() != 1) {
+                throw lsst::partition::ConfigTypeError(
+                    context + "parameter: '" + path + "' is a string, but not the single-character one"
+                    " to allow interpreting it as a value of the 'char' type.");
+            }
+            return s[0];
+        }
+        try {
+            return param.get<char>();
+        } catch(std::exception const& ex) {
+            throw lsst::partition::ConfigTypeError(
+                context + "incorrect type conversion for parameter: '" + path + "', details: "
+                + std::string(ex.what()));
+        }
+    }
+};
+}}} // namespace lsst::partition::detail
+
+
+namespace lsst {
+namespace partition {
+/**
+ * Class ConfigStore is a unified transient storage of the configuration parameters read
+ * from the configuration files or command-line parameters.
+ * 
+ * Parameters are fetched from the store using their "path" specification. The dot
+ * character '.' is used for separating components of the path:
+ * @code
+ * foo.bar.p1
+ * dir.p2
+ * p3
+ * @code
+ * The intermediate elements of the path would directly map to the corresponding keys
+ * of the nested JSON objects in the store. For example, the above shown paths would
+ * be represented by the following JSON structure passed into the class's constructr,
+ * into method add(), or read from an input file when calling method parse().
+ * from the :
+ * @code
+ * {
+ *   "foo": {
+ *       "bar": {
+ *           "p1": <value>
+ *       }
+ *    },
+ *    "dir": {
+ *       "p2": <value>
+ *    },
+ *    "p3": <value>
+ * }
+ * @code
+ * Values of the command-line parameters are expected to be injected into the configuration
+ * by calling method set(). The long names (excluding preceeding "--") of the parameters
+ * would be treated as above explained path names.
+ *
+ * @note the class's implementation is not thread-safe, and any subsequent modification
+ *   to a parameter would always override its previous value (if any existed).
```

Comment:
I think these are unrelated points? 
1. It is not thread-safe (unprotected access could crash)
2. Modifications are last-one wins. 
If so, it probably deserves separate sentences for each point, for clarity.